server/src/main/scala/org/scassandra/server/priming/cors/CorsSupport.scala

```diff
@@ -0,0 +1,43 @@
+package org.scassandra.server.priming.cors
+
+import com.typesafe.scalalogging.LazyLogging
+import org.scassandra.server.ScassandraConfig
+import spray.http.HttpHeaders._
+import spray.http.HttpMethods._
+import spray.http.{AllOrigins, HttpMethods, HttpResponse}
+import spray.routing._
+
+/**
+  * A mixin to provide support for providing CORS headers as appropriate
+  */
+trait CorsSupport {
+  this: HttpService with LazyLogging =>
+
+  private val allowOriginHeader = `Access-Control-Allow-Origin`(AllOrigins)
+  private val optionsCorsHeaders = List(
+    `Access-Control-Allow-Headers`(
+      "Origin, X-Requested-With, Content-Type, Accept, Accept-Encoding, Accept-Language, Host, Referer, User-Agent"
+    ),
+    `Access-Control-Max-Age`(60 * 60 * 24 * 20) // cache pre-flight response for 20 days
+  )
+
+  def cors[T]: Directive0 = mapRequestContext {
+    context => context.withRouteResponseHandling {
+      // If an OPTIONS request was rejected as 405, complete the request by responding with the
+      // defined CORS details and the allowed options grabbed from the rejection
+      case Rejected(reasons) if
```

Comment:
I guess you already checked this, but just to clarify.
When you use a client library like jQuery or Axios and you specify that's a CORS request, 2 things happens on every request:

- First, a preflight request is sent with HTTP Verb OPTIONS for a given endpoint.
  Then, if the server supports CORS, it replies back saying what verbs, headers and domains that endpoint supports.

- Secondly, the client requesting the resource based on the preflight request response decides to proceed or not.
  Typically, if the preflight fails there makes no sense to proceed and the client fails gracefully with a CORS error message.

In the current implementation we are handling a 405 / Method Not Allowed and if we are in the presence of a OPTIONS (preflight) request, 
we just send back the appropriated headers required for CORS.


> On 31 May 2016, at 13:19, Andrew Tolbert <notifications@github.com> wrote:
> 
> In server/src/main/scala/org/scassandra/server/priming/cors/CorsSupport.scala <https://github.com/scassandra/scassandra-server/pull/162#discussion_r65170457>:
> 
> > +trait CorsSupport {
> > +  this: HttpService with LazyLogging =>
> > +
> > +  private val allowOriginHeader = `Access-Control-Allow-Origin`(AllOrigins)
> > +  private val optionsCorsHeaders = List(
> > +    `Access-Control-Allow-Headers`(
> > +      "Origin, X-Requested-With, Content-Type, Accept, Accept-Encoding, Accept-Language, Host, Referer, User-Agent"
> > +    ),
> > +    `Access-Control-Max-Age`(60 * 60 * 24 * 20) // cache pre-flight response for 20 days
> > +  )
> > +
> > +  def cors[T]: Directive0 = mapRequestContext {
> > +    context => context.withRouteResponseHandling {
> > +      // If an OPTIONS request was rejected as 405, complete the request by responding with the
> > +      // defined CORS details and the allowed options grabbed from the rejection
> > +      case Rejected(reasons) if
> I'm not entirely sure how this works since I haven't used spray that much, is this intercepting what would be a 405 response and injecting the appropriate response with all the headers?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/scassandra/scassandra-server/pull/162/files/25f99748f09f8a6290c5b72e805bde6764026c8e#r65170457>, or mute the thread <https://github.com/notifications/unsubscribe/AEcH5yqY0Qy1LQjcnX4gVDeN4vBeIOCgks5qHCc2gaJpZM4Io98s>.
>